registry/cache/cache.go

```diff
@@ -101,38 +100,44 @@ func (c *cache) quit() bool {
 	}
 }
 
-func (c *cache) del(service string) {
+func (c *cache) del(domain, service string) {
 	// don't blow away cache in error state
 	if err := c.status; err != nil {
 		return
 	}
-	// otherwise delete entries
-	delete(c.cache, service)
-	delete(c.ttls, service)
+
+	if _, ok := c.services[domain]; ok {
+		delete(c.services[domain], service)
+	}
+
+	if _, ok := c.ttls[domain]; ok {
+		delete(c.ttls, service)
+	}
 }
 
-func (c *cache) get(service string) ([]*registry.Service, error) {
-	// read lock
-	c.RLock()
+func (c *cache) get(domain, service string) ([]*registry.Service, error) {
+	var services []*registry.Service
+	var ttl time.Time
 
-	// check the cache first
-	services := c.cache[service]
-	// get cache ttl
-	ttl := c.ttls[service]
-	// make a copy
-	cp := util.Copy(services)
+	// lookup the values in the cache before calling the underlying registrry
+	c.RLock()
+	if srvs, ok := c.services[domain]; ok {
```

Comment:
Some accesses to c.services are done under lock but others aren't (see `del()`). Needs to be consistent

Reply:
Actioned